src/test/SubscribableRequestQueue.test.ts

```diff
@@ -0,0 +1,262 @@
+
+/** Returns a promise that resolves after `timeoutMs` milliseconds with value `result`. */
+function delay<T>(timeoutMs: number, result: T): Promise<T> {
+  return new Promise((resolve) => setTimeout(() => resolve(result), timeoutMs));
+}
+
+/** Returns a promise that rejects after `timeoutMs` milliseconds with reason `reason`. */
+function delayReject(timeoutMs: number, reason: unknown): Promise<unknown> {
+  return new Promise((_, reject) => setTimeout(() => reject(reason), timeoutMs));
+}
+
+/** Determines whether a promise has rejected. */
+// From https://gist.github.com/tyru/29360dfa475d2fefaf6c4655a93c2cb0
+function isRejected(promise: Promise<unknown>): Promise<boolean> {
+  return Promise.race([
+    delay(0, false),
+    promise.then(
+      () => false,
+      () => true
+    ),
+  ]);
+}
+
+describe("SubscribableRequestQueue", () => {
+  describe("addSubscriber", () => {
+    it("adds a subscriber", () => {
+      const queue = new SubscribableRequestQueue();
+      const id = queue.addSubscriber();
+      expect(queue.hasSubscriber(id)).to.be.true;
+    });
+
+    it("returns unique ids", () => {
+      const queue = new SubscribableRequestQueue();
+      const id1 = queue.addSubscriber();
+      const id2 = queue.addSubscriber();
+      expect(id1).to.not.equal(id2);
+    });
+  });
+
+  describe("addRequestToQueue", () => {
+    it("queues a request", async () => {
+      const queue = new SubscribableRequestQueue();
+      const id = queue.addSubscriber();
+      const promise = queue.addRequestToQueue("test", id, () => delay(TIMEOUT, "foo"));
+      expect(queue.hasRequest("test")).to.be.true;
+      const result = await promise;
+      expect(result).to.equal("foo");
+    });
+
+    it("does not queue a request if the subscriber id is invalid", () => {
+      const queue = new SubscribableRequestQueue();
+      expect(() => queue.addRequestToQueue("test", -1, () => delay(TIMEOUT, "foo"))).to.throw();
+      expect(() => queue.addRequestToQueue("test", 0, () => delay(TIMEOUT, "foo"))).to.throw();
+    });
+
+    it("does not queue a request if the subscriber has been deleted", () => {
+      const queue = new SubscribableRequestQueue();
+      const id = queue.addSubscriber();
+      queue.removeSubscriber(id);
+      expect(() => queue.addRequestToQueue("test", id, () => delay(TIMEOUT, "foo"))).to.throw();
+    });
+
+    it("creates unique promises without duplicating work when two subscribers queue the same key", async () => {
+      const queue = new SubscribableRequestQueue();
+      const id1 = queue.addSubscriber();
+      const id2 = queue.addSubscriber();
+
+      const promise1 = queue.addRequestToQueue("test", id1, () => delay(TIMEOUT, "foo"));
+      const promise2 = queue.addRequestToQueue("test", id2, () => delay(TIMEOUT, "bar"));
+      expect(queue.hasRequest("test")).to.be.true;
+      expect(promise1).to.not.equal(promise2);
+
+      const [result1, result2] = await Promise.all([promise1, promise2]);
```

Comment:
this test makes me think that subscriber 1 wants to do the "foo" result and subscriber2 wants "bar" when the same data request arrives.  Can two subscribers use the same key and do different things when the request completes?

Reply:
I think my confusion was in forgetting that the last arg is the whole process and not the "then" part.   It seems odd to add requests consecutively in this way, when they don't know about each other and are passing in totally different functions.

Reply:
Right, the "bar" task never starts executing. The queue assumes that clients will ensure that requests with the same key, even when initiated by different subscribers, do effectively the same thing. For example, a request keyed by a specific chunk URL will always be a request to fetch that chunk. Given that assumption, it makes sense for a new subscription to an existing key to join the work in progress rather than interrupt it and start over. That's what this test is about.

Reply:
Okay.  So my problem is that the api is weird because the caller is expected to pass in what might be a totally unique task function and it's possible that it will never ever run.  How can the caller receive that information (that their task isn't going to happen because someone else already submitted a task for this)?  (Or should the caller just not care?)